### DISCUSSION TITLE
Range selectors for ligatures does not work properly with aat fonts applying ligatures via morx table

### DISCUSSION THREAD

**Discussion by DeependraKhetwal** [harfbuzz/harfbuzz#4019] [General] [answered]
Hi , I am facing an issue with aat fonts where on disabling ligatures for a particular range the ligatures gets disabled for the whole line. The issue does not happen with fonts that apply ligature substitutions via opentype 'liga' feature but which apply ligature substitutions via 'morx' table. In the attached example I am disabling ligatures for 'ffi' (using -liga[3:])  but ligature is getting turned off for 'ff' also.
<img width="571" alt="Screenshot 2022-12-05 at 2 02 42 PM (1)" src="https://user-images.githubusercontent.com/122281418/211316367-7a7ceb8f-e6cf-4078-937d-650481fad787.png">

**Comment by behdad** [comment, marked as answer]
You are correct. We don't quite know how to implement that. I need to test if CoreText implements it at all.

**Reply by jfkthame** [reply]
I would assume -- but have not tested -- that you could do it via functions like CTLineCreateWithAttributedString, by setting different CTFont instances with appropriate features for the different ranges of the string.

(Or in simple cases, maybe just setting different kCTLigatureAttributeName values for ranges of the attributed string, but that doesn't look like it'd support arbitrary features/settings, only a few "well-known" options.)

**Reply by behdad** [reply]
> I would assume -- but have not tested -- that you could do it via functions like CTLineCreateWithAttributedString, by setting different CTFont instances with appropriate features for the different ranges of the string.

We do that. I'm not sure how to implement it using the state-machine though.

**Reply by behdad** [reply]
> We do that. I'm not sure how to implement it using the state-machine though.

Maybe do the state transition but not apply the action. I'll experiment with that.

**Reply by behdad** [reply]
Or maybe it restarts the machine on feature changes. That makes the most sense. That would require some drastic changes to how we do things.

**Reply by behdad** [reply]
> Or maybe it restarts the machine on feature changes. That makes the most sense. That would require some drastic changes to how we do things.

@nedley can you advise how CoreText deals with features (liga, etc) changing mid-string? Does the state machine restart from the beginning?

**Reply by nedley** [reply]
At the very least adjacent runs with different feature settings need to be treated as independent runs, otherwise you can decide whether to break shaping for a subtable if it has different resolved flags between the runs.

**Reply by behdad** [reply]
Thanks @nedley.

@DeependraKhetwal I created an issue to track implementing this. https://github.com/harfbuzz/harfbuzz/issues/4020

**Reply by DeependraKhetwal** [reply]
Thanks @behdad
